[JSC] Use immediate bit-vectors for character class matching in YarrJIT

ddegazio opened the pull request:
#### 34b0b047bb64f93ccd1b003d410e0f8b4c9d681b
<pre>
[JSC] Use immediate bit-vectors for character class matching in YarrJIT
<a href="https://bugs.webkit.org/show_bug.cgi?id=275279">https://bugs.webkit.org/show_bug.cgi?id=275279</a>
<a href="https://rdar.apple.com/129419939">rdar://129419939</a>

Reviewed by Michael Saboff.

Changes how YarrJIT handles character class matches via the following:

 1. Optimize single-range checks from two branches into subtract + branch.

 2. Use a bit-vector test to quickly match a set of individual characters,
    as opposed to the current strategy of O(n) sequential equality checks.

 3. Make the logic of matchCharacterClassRange more recursive. We use the
    optimized single-range test if there is only a single range, and use
    the new bit-vector test if the whole set of ranges and character matches
    fits within a small-enough range. Moreover, the binary search is now
    totally recursive, meaning we can use these specialized checks for
    recursive checks within the binary search too, whereas currently binary
    search is kind of all-or-nothing.

 4. A few small optimizations are removed - YarrJIT no longer special-cases
    ASCII letters in character class matches, since character set matching
    is now faster. Turning adjacent character matches into length-two ranges
    is also removed during CharacterClass construction since this doesn&apos;t
    really do anything other than make the binary search do extra work (I&apos;d
    be really surprised if this was ever particularly profitable).

Overall, this seems to be a somewhat modest but appreciable perf win on
microbenchmarks. On the added ASCII alphanumeric test I&apos;m seeing about 10%
improvement with this new approach, and on the single-chars test I&apos;m seeing
more like 20% improvement. I&apos;ve added a test for a set of separator chars
too, and we have maybe a small ~2% improvement on my machine - this is pretty
small and hopefully improvable? Not so exciting, but let&apos;s have the
microbenchmark in the tree anyway.

* JSTests/microbenchmarks/regexp-match-alphanumeric.js: Added.
* JSTests/microbenchmarks/regexp-match-multiple-single-chars.js: Added.
* JSTests/microbenchmarks/regexp-match-separators.js: Added.
(let.src):
(dot):
(test):
(i.let.re):
* Source/JavaScriptCore/yarr/YarrJIT.cpp:
* Source/JavaScriptCore/yarr/YarrPattern.cpp:
(JSC::Yarr::CharacterClassConstructor::addSorted):

Canonical link: <a href="https://commits.webkit.org/280425@main">https://commits.webkit.org/280425@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0a27cdbc29c374e37fa153edd511fc016608f7b5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56507 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35833 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8979 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60114 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6950 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43455 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7139 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45783 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4865 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58536 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33708 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48777 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26636 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30471 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6104 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/5947 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/49599 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/52451 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6363 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/61798 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/55746 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/415 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6487 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53043 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/416 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48827 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/52950 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12527 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/370 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/77509 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31661 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12844 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32746 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33837 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32494 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/yarr/YarrJIT.cpp:
```diff
@@ -523,62 +524,198 @@ class YarrGenerator final : public YarrJITInfo {
+        switch (mask.count()) {
+        case 0:
+            return;
+        case 1 ... 3:
```
Let's list up all of integers since this is GCC extension.

ddegazio replied:
Done in latest push.

Constellation commented on Source/JavaScriptCore/yarr/YarrJIT.cpp:
```diff
@@ -523,62 +524,198 @@ class YarrGenerator final : public YarrJITInfo {
+            break;
+        default: {
+            // Otherwise, actually perform the bit test.
+            AllowMacroScratchRegisterUsage scope(m_jit);
```
Is it allowed in YarrJIT?

msaboff replied:
The YarrJIT indirectly uses scratch registers via the macro assembler.  A couple of examples are address calculations and (I believe) materializing backtrack code addresses.

ddegazio replied:
It's allowed, but lshift32/64 use the scratch register on x86/armv7, so we clobber it...I've moved this logic inside of the macro assembler in the latest push.

aoikonomopoulos commented on Source/JavaScriptCore/assembler/MacroAssemblerARMv7.h:
```diff
@@ -386,6 +386,14 @@ class MacroAssemblerARMv7 : public AbstractMacroAssembler<Assembler> {
+        // Clamp the shift to the range 0..31
+        m_assembler.ARM_and(dest, shiftAmount, ARMThumbImmediate::makeEncodedImm(0x1f));
+        move(imm, getCachedAddressTempRegisterIDAndInvalidate());
+        m_assembler.lsl(dest, dataTempRegister, dest);
```
Hmm, the `lsl` used `dataTempRegister` but the immediately preceding `move` defines `addressTempRegister` instead?

ddegazio replied:
That sounds like it could be it! I'll give it a try.

msaboff commented on Source/JavaScriptCore/yarr/YarrPattern.cpp:
```diff
@@ -481,31 +481,9 @@ class CharacterClassConstructor {
-                    addSortedRange(isASCII(ch) ? m_ranges : m_rangesUnicode, lo, hi);
-                    return;
-                }
+            else if (val > 0)
```
I can't remember why I did this coalescing while adding individual characters, but I seem to remember there were pathological cases where some ops on large character classes became unwieldy, i.e. large vectors.  That might have been before I implemented the operations with chunked bitmaps.  Certainly, `coalesceTables()` does the coalescing we need before returning the created character class.

ddegazio replied:
Having more pairs pessimized a few character class matches before, but I think this is pretty niche (it can only happen when one of these pairs would cross a set boundary, not sure it actually really matters outside of specific examples. I'll revert this, we don't need to overfit.

msaboff reviewed (approved):
r=me

Do you have any plans to expand the bit testing to Unicode characters?  Even using `matchCharacterClassOnlyOneRange()` would be helpful for unicode character classes that are alphabets.

ddegazio committed:
[JSC] Use immediate bit-vectors for character class matching in YarrJIT
https://bugs.webkit.org/show_bug.cgi?id=275279
rdar://129419939

Reviewed by Michael Saboff.

Changes how YarrJIT handles character class matches via the following:

 1. Optimize single-range checks from two branches into subtract + branch.

 2. Use a bit-vector test to quickly match a set of individual characters,
    as opposed to the current strategy of O(n) sequential equality checks.

 3. Make the logic of matchCharacterClassRange more recursive. We use the
    optimized single-range test if there is only a single range, and use
    the new bit-vector test if the whole set of ranges and character matches
    fits within a small-enough range. Moreover, the binary search is now
    totally recursive, meaning we can use these specialized checks for
    recursive checks within the binary search too, whereas currently binary
    search is kind of all-or-nothing.

 4. A few small optimizations are removed - YarrJIT no longer special-cases
    ASCII letters in character class matches, since character set matching
    is now faster. Turning adjacent character matches into length-two ranges
    is also removed during CharacterClass construction since this doesn't
    really do anything other than make the binary search do extra work (I'd
    be really surprised if this was ever particularly profitable).

Overall, this seems to be a somewhat modest but appreciable perf win on
microbenchmarks. On the added ASCII alphanumeric test I'm seeing about 10%
improvement with this new approach, and on the single-chars test I'm seeing
more like 20% improvement. I've added a test for a set of separator chars
too, and we have maybe a small ~2% improvement on my machine - this is pretty
small and hopefully improvable? Not so exciting, but let's have the
microbenchmark in the tree anyway.

* JSTests/microbenchmarks/regexp-match-alphanumeric.js: Added.
* JSTests/microbenchmarks/regexp-match-multiple-single-chars.js: Added.
* JSTests/microbenchmarks/regexp-match-separators.js: Added.
(let.src):
(dot):
(test):
(i.let.re):
* Source/JavaScriptCore/yarr/YarrJIT.cpp:
* Source/JavaScriptCore/yarr/YarrPattern.cpp:
(JSC::Yarr::CharacterClassConstructor::addSorted):

Canonical link: https://commits.webkit.org/280425@main
